### PR TITLE
[WIP] Port MONAI Generative utils

### PR DESCRIPTION
**Pull request by marksgraham**
Towards completing #6676 .

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by marksgraham** [comment]
@ericspod I'm getting mypy errors from `unsqueeze_left` and `unsqueeze_right`:

```
mypy 1.4.1 (compiled: yes)
monai/utils/misc.py:895:5: error: Returning Any from function declared to return "T"  [no-any-return]
monai/utils/misc.py:895:12: error: Value of type "object" is not indexable  [index]
monai/utils/misc.py:895:43: error: "T" has no attribute "ndim"  [attr-defined]
monai/utils/misc.py:900:5: error: Returning Any from function declared to return "T"  [no-any-return]
monai/utils/misc.py:900:12: error: Value of type "object" is not indexable  [index]
monai/utils/misc.py:900:34: error: "T" has no attribute "ndim"  [attr-defined]
```

so far they only operate on tensors - am I ok to type them specifally for tensors/np arrays to satisfy mypy?

**Comment by ericspod** [comment]
> am I ok to type them specifally for tensors/np arrays to satisfy mypy?

If you need to yes do that, but we do have [NdarrayOrTensor](https://github.com/Project-MONAI/MONAI/blob/0a3cdd9f1403cbad057b96230840ee17cf430dbf/monai/config/type_definitions.py#L69).

File: monai/utils/enums.py [hunk 1]
```diff
@@ -692,3 +703,39 @@ class AlgoKeys(StrEnum):
     ALGO = "algo_instance"
     IS_TRAINED = "is_trained"
     SCORE = "best_metric"
+
+
+class AdversarialKeys(StrEnum):
```
Comment (KumoLiu) [inline monai/utils/enums.py:708]: Would you mind adding a brief description here and also for `AdversarialIterationEvents`?

File: tests/test_squeeze_unsqueeze.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import unittest
+
+import numpy as np
+import torch
+from parameterized import parameterized
+
+from monai.utils import unsqueeze_left, unsqueeze_right
+
+RIGHT_CASES = [(np.random.rand(3, 4), 5, (3, 4, 1, 1, 1)), (torch.rand(3, 4), 5, (3, 4, 1, 1, 1))]
```
Comment (KumoLiu) [inline tests/test_squeeze_unsqueeze.py:22]: Could you please also test other data types here?
An example here:
https://github.com/Project-MONAI/MONAI/blob/0a3cdd9f1403cbad057b96230840ee17cf430dbf/tests/test_normalize_intensity.py#L24

**Review by KumoLiu — APPROVED** [review]
Thanks for the PR, few minor comments inline.

File: monai/utils/component_store.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+T = TypeVar("T")
+
+
+def is_variable(name):
+    """Returns True if `name` is a valid Python variable name and also not a keyword."""
+    return name.isidentifier() and not iskeyword(name)
+
+
+class ComponentStore:
```
Comment (wyli) [inline monai/utils/component_store.py:27]: we might be able to use/enhance the existing `LayerFactory` to achieve the same capability https://github.com/Project-MONAI/MONAI/blob/0a3cdd9f1403cbad057b96230840ee17cf430dbf/monai/networks/layers/factories.py#L76 

cc @ericspod I think the usage is https://github.com/Project-MONAI/GenerativeModels/blob/f28afa003f9b5be0ebf9f31270abc06e7957a607/generative/networks/schedulers/scheduler.py#L44

Comment (ericspod) [reply monai/utils/component_store.py:27]: > we might be able to use/enhance the existing `LayerFactory` to achieve the same capability

I thought about this before when writing this initially but concluded that there were too many differences between the two classes. We could refactor things such that they inherit from a common implementation, we need two classes because the usage here isn't to generate layers but be a factory for functions.

Comment (ericspod) [reply monai/utils/component_store.py:27]: But again these aren't network layers. The purpose of LayerFactory is to be a factory for the components of a network, so at the least we need to rename classes a bit.

Comment (wyli) [reply monai/utils/component_store.py:27]: indeed, the `LayerFactory` doesn't assume network components/blocks, the class could be renamed to a more generic one and deprecating the existing one. or have a refactoring into two subclasses as @ericspod mentioned. would you be interested in implementing it? @marksgraham

Comment (marksgraham) [reply monai/utils/component_store.py:27]: so a `Factory` class which `LayerFactory` and `ComponentStore` inherit from? That to me seems preferable to deprecating `LayerFactory`.

Comment (wyli) [reply monai/utils/component_store.py:27]: sounds good, thanks @marksgraham!

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me.
